### DISCUSSION TITLE
Wake host display(s) on connect instead of erroring out

### DISCUSSION THREAD

**Discussion by drws** [LizardByte/community#439] [Sunshine]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

Sunshine currently fails to initialize a session when host display is turned off. This is easy to solve for a great majority of cases by issuing a DPMS turn on command such as `xset dpms force on`. This could and should be automatic and built-in, possibly configurable (if there are use cases) in Sunshine and/or Moonlight.

### Expected Behavior

Sunshine should know the simple operation of turning on display instead of relying on the user to manually ensure for the display to be turned on when a remote session is wanted. Especially since Moonlight also supports Wake-on-LAN and turning on display seems an even more frequent case

### Host Operating System

Linux

### Operating System Version

Debian 12

### Architecture

64 bit

### Sunshine commit or version

0.23.1

### Package

Linux - deb

### GPU Type

none (software encoding)

### GPU Model

Intel

### GPU Driver/Mesa Version

i915

### Config

```Shell
Default configuration.
```

### Relevant log output

Moonlight reports the following error on connection try:

```shell
Host returned error. Failed to initialize video capture/encoding. Is a display connected and turned on? (Error 503)
```

**Comment by vvuk** [comment]
This would be great; took me a while to figure out what was happening and also how to recover. In my case at least (i915), I have to both `xset dpms force on` and restart the sunshine service. Then I'm able to connect. So beyond just kicking a monitor on, sunshine may need to also reinitialize capture.

**Comment by t184256** [comment]
What's even worse, I can't seem to bind waking up the display to a global_prep_cmd or an app one, because it errors out before triggering the command =(

**Comment by Damglador** [comment]
Is there a way to kick display on with on Plasma Wayland?

**Reply by treierxyz** [reply]
If you have SSH access, you can run `qdbus org.kde.Solid.PowerManagement /org/kde/Solid/PowerManagement org.kde.Solid.PowerManagement.wakeup` on the Sunshine host machine to wake up Plasma Wayland. As @t184256 mentioned, you can't add this to the Sunshine prep commands as they don't run before the error occurs.

**Reply by Damglador** [reply]
Got ssh working, command works perfectly. Now time to figure out how to make a script for it and add it on my launcher with Termux. Thank you very much🤝

**Reply by Tebro** [reply]
With KDE you can use `kscreen-doctor --dpms on` to turn screens on, and if you do it over SSH you need the extra env variable (assuming Wayland) `QT_QPA_PLATFORM=wayland kscreen-doctor`

**Comment by t184256** [comment]
Just wiggle the mouse already. Sunshine creates the device anyway, how hard would it be?

**Reply by Damglador** [reply]
Well, how to wiggle a mouse if I can't connect to the system with Moonlight?

I was thinking about using KDE Connect, but it doesn't work through ZeroTier

**Reply by ReenigneArcher** [reply]
I believe they are suggesting that Sunshine should "wiggle the mouse".

**Comment by t184256** [comment]
I just wrote a tiny Python uinput script that monitors logs and wiggles when it sees errors: https://github.com/t184256/nix-configs/blob/58c4fd303a12a69f33f99f72f7263e1b2b5b8e92/nixos/services/sunshine.nix#L29

But I was addressing sunshine developers, really. Sunshine has all it needs (virtual mouse device included), but the will to do the wiggle.

**Reply by ReenigneArcher** [reply]
Feel free to submit a PR. The very few of us actually working on Sunshine don't have time to work on everyone's requests.

**Reply by drws** [reply]
And in such case please use proper calls, not actual wiggling :)

**Comment by OhaDerErste** [comment]
Any news on this?

**Comment by samurailink3** [comment]
I also ran into this issue, so I wrote a little Golang service to watch the logs for monitor errors, wake the monitor, then restart Sunshine. This seems to be working fine for me (although a bit annoying compared to a native solution). Hopefully this helps someone else: https://github.com/samurailink3/sunrise

**Reply by Quadrapole** [reply]
thanks for this. good work around for now

**Comment by l-raider** [comment]
Both "kscreen-doctor --dpms on" and "qdbus org.kde.Solid.PowerManagement /org/kde/Solid/PowerManagement org.kde.Solid.PowerManagement.wakeup" work when ran manually, any idea why they don't work as Sunshine pre commands?

Are the pre-commands only executed IF there is a display found?

**Reply by Tebro** [reply]
That is my understanding, we need to get these commands in earlier in the flow to have the screen turn on earlier. Something like early_pre_commands could be interesting.

**Comment by Tebro** [comment]
As a feature request: I would like to suggest that we could have some sort of pre_stream_commands that can be configured like the prep commands, but that would be run earlier. This would allow us to setup commands that can turn the display on as you connect to Sunshine. 

I explored the code with Codex and it should be doable as a new configuration section, but I am no C++ developer so I cannot really review how good of a solution it is that Codex has suggested.

**Reply by ReenigneArcher** [reply]
If you're interested in this please leave a thumbs up on https://github.com/LizardByte/roadmap/issues/38

**Comment by hubbaredneck** [comment]
Workaround systemd service that runs kscreendoctor when a new TCP connection is detected on sunshines port for nixos:

```
  # conntrack needs cap_net_admin to run as a normal user
  security.wrappers.conntrack = {
    source = "${pkgs.conntrack-tools}/bin/conntrack";
    capabilities = "cap_net_admin+ep";
    owner = "root";
    group = "root";
    permissions = "0755";
  };

  # sunshine wake display script
  environment.etc."sunshine_wake_display.sh".text = ''
    #!${pkgs.bash}/bin/bash
    /run/wrappers/bin/conntrack -E -e new -p tcp --dport 47984 | \
      while read line; do
        /run/current-system/sw/bin/logger -s "New Sunshine connection on port 47984, running kscreen-doctor -d on"
        /run/current-system/sw/bin/kscreen-doctor -d on
        sleep 5
      done
  '';

  # systemd service to wakeup monitor on sunshine connection
  systemd.user.services.sunshine-monitor = {
    description = "Monitor Sunshine TCP connections and enable display";

    wantedBy = [ "multi-user.target" ];
    after = [ "network-online.target" ];
    wants = [ "network-online.target" ];

    serviceConfig = {
      Type = "simple";
      ExecStart = "${pkgs.bash}/bin/bash /etc/sunshine_wake_display.sh";
      Restart = "always";
      RestartSec = 3;

      Environment = "PATH=/run/current-system/sw/bin";
    };
  };

```

**Reply by magic0whi** [reply]
> Workaround systemd service that runs kscreendoctor when a new TCP connection is detected on sunshines port for nixos:
> 
> ```
>   # conntrack needs cap_net_admin to run as a normal user
>   security.wrappers.conntrack = {
>     source = "${pkgs.conntrack-tools}/bin/conntrack";
>     capabilities = "cap_net_admin+ep";
>     owner = "root";
>     group = "root";
>     permissions = "0755";
>   };
> 
>   # sunshine wake display script
>   environment.etc."sunshine_wake_display.sh".text = ''
>     #!${pkgs.bash}/bin/bash
>     /run/wrappers/bin/conntrack -E -e new -p tcp --dport 47984 | \
>       while read line; do
>         /run/current-system/sw/bin/logger -s "New Sunshine connection on port 47984, running kscreen-doctor -d on"
>         /run/current-system/sw/bin/kscreen-doctor -d on
>         sleep 5
>       done
>   '';
> 
>   # systemd service to wakeup monitor on sunshine connection
>   systemd.user.services.sunshine-monitor = {
>     description = "Monitor Sunshine TCP connections and enable display";
> 
>     wantedBy = [ "multi-user.target" ];
>     after = [ "network-online.target" ];
>     wants = [ "network-online.target" ];
> 
>     serviceConfig = {
>       Type = "simple";
>       ExecStart = "${pkgs.bash}/bin/bash /etc/sunshine_wake_display.sh";
>       Restart = "always";
>       RestartSec = 3;
> 
>       Environment = "PATH=/run/current-system/sw/bin";
>     };
>   };
> ```

Thanks for the workaround! Here is the configuration adapted for Hyprland. I also improved in a more Nix-way:
```nix
  # https://github.com/orgs/LizardByte/discussions/439#discussioncomment-15813284
  security.wrappers.conntrack = lib.mkIf config.services.sunshine.enable {
    source = "${pkgs.conntrack-tools}/bin/conntrack";
    # conntrack needs cap_net_admin to run as a normal user
    capabilities = "cap_net_admin+ep";
    owner = "root"; group = "root";
  };
  systemd.user.services.sunshine-wake-monitor= lib.mkIf config.services.sunshine.enable {
    description = "Monitor Sunshine TCP connections and wake monitors";
    after = ["hyprland-session.target"];
    wantedBy = ["graphical-session.target"];
    serviceConfig = {
      ExecStart = pkgs.writeShellScript "sunshine_wake_monitor" ''
        ${config.security.wrapperDir}/conntrack -E -e new -p tcp --dport ${builtins.toString (config.services.sunshine.settings.port - 5)} | \
        while read line; do
          echo "New Sunshine connection detected, waking up the monitors"
          ${lib.getExe' pkgs.hyprland "hyprctl"} --instance 0 'dispatch dpms on'
          sleep 5
        done
      '';
      Restart = "on-failure";
    };
  };
```

**Comment by Hely0n** [comment]
I found the following workaround for KDE Plasma:

1. Install knockd with your package manager (it's just 100kb 😉)
2. Make your /etc/knockd.conf file look like this:

```
[options]
	UseSysLog
	Interface = <YOUR INTERFACE>

[WakeOnMoonlight]
    sequence    = 47989
    seq_timeout = 5
    tcpflags    = syn
    command     = XDG_RUNTIME_DIR=/run/user/1000 WAYLAND_DISPLAY=wayland-0 kscreen-doctor --dpms on
```
    
3. `sudo systemctl enable --now knockd.service`

This should already wake your monitors when you open the moonlight app 😄 
I'm on CachyOS (Arch based) but I think it should work on other distros with KDE Plasma too.

It's meant to work with Wayland, not X11

Edit:

I added a debounce of 1 minute to the command, so that kscreen-doctor won't be spammed while moonlight is open:

```
command     = flock -n /tmp/kscreen.lock -c 'XDG_RUNTIME_DIR=/run/user/1000 WAYLAND_DISPLAY=wayland-0 kscreen-doctor --dpms on; sleep 60```

**Reply by Hely0n** [reply]
PS: I had the problem that it didn't work when knockd was started on boot. If someone encounters the same problem, the fix was to add 
```
[Unit]
After=network-online.target
Wants=network-online.target
```

to the knockd.service

**Reply by hubbaredneck** [reply]
Nice idea to use knockd, this is probably more reliable than my conntrack script. Will try to implement it for nixos later.

**Reply by Hely0n** [reply]
Thank you. When I saw your workaround, I actually thought conntrack is something preinstalled and a cleaner solution than mine 😄

**Reply by ewgdg** [reply]
In my case, I found that Moonlight periodically sends SYN packets to ports 47984 and 47989 for health checking, which unnecessarily keeps the display awake when Moonlight is open. This is not ideal.
I guess we are still waiting for Sunshine to implement pre_stream_commands.

**Reply by Hely0n** [reply]
yes, unfortunately this workaround only applies to users who occationally open moonlight to stream. It's no replacement for a pre_stream_commands feature

**Reply by MartenBE** [reply]
> Nice idea to use knockd, this is probably more reliable than my conntrack script. Will try to implement it for nixos later.

Unfortunately, it seems that knockd is not available in the Nix repo :/

**Comment by marcusnewton** [comment]
This should really be a built-in feature for Sunshine to wake the display automatically.
